### ISSUE TITLE
Any key press elicits a configured phrase/script, bug in "/usr/lib/python3.6/site-packages/autokey/service.py"

### ISSUE THREAD

**Issue by CCmesch** [autokey/autokey#247]
## Classification:

Bug

## Reproducibility:

 Always

## Version

AutoKey version: 

Autokey 0.59.4
Qt 5.9.4
(Window manager xfce4)

Installed via: package manager in opensuse (zypper)

Linux Distribution: Opensuse Leap 15.0

## Summary

When I log into the computer and start autokey (monitoring enabled) all keypresses elicit one of my configured phrases or sample scripts, although there are defined abbreviations, which are disregarded. EACH keypress, no matter which key, gives out a phrase/script until I stop monitoring. I then have to stop monitoring in order to do some work.  The autokey log shows suspicious error items concerning  "__tryReleaseLock":

-------------------------------- autokey log  ----------------------------------------
```
Raw key: 'j', modifiers: [], Key: j
Window visible title: 'Terminal - me@computer:~', Window class: 'xfce4-terminal.Xfce4-terminal'
No phrase/script matched hotkey
Send via event interface
Send special key: [<Key.BACKSPACE: '<backspace>'>]
Input stack at end of handle_keypress: deque([], maxlen=150)
Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/autokey/service.py", line 206, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
Send special key: [<Key.BACKSPACE: '<backspace>'>]
Send special key: [<Key.BACKSPACE: '<backspace>'>]
Send special key: [<Key.BACKSPACE: '<backspace>'>]
Send special key: [<Key.BACKSPACE: '<backspace>'>]
Sending string: 'my configured phrase here'
```
## Steps to Reproduce (if applicable)

start monitoring ; press any key; see wrong output

## Expected Results

normally "my phrase" should only be triggered on the configured abbreviation (e.g. +myphrase )
I don't use hotkeys

## Actual Results

- Instead, on any key press "myphrase" or "mysymplescript" is triggered

**Comment by luziferius** [comment]
This certainly looks odd. And you meant version 0.95.4, right?

I suspect that the meta-data is somehow corrupted, causing the checks to always succeed and thus immediately trigger your phrase.

Would you mind uploading the meta-data for the offending phrase?

In the AutoKey GUI, select the folder your phrase is in. In the top right (above **`Folder Settings`**, there is a link to the folder on disk, which opens your file manager there. I’m interested in the file `.myphrase.json`(it is named the same as your phrase), which is normally a hidden file.

It contains the exact phrase configuration, like the abbreviation, window filter and all other settings for the given phrase. Please ZIP archive it (GitHub disallows `.json` files) and attach it here.

I’ll take a look at it.

**Comment by josephj11** [comment]
In addition to what @luziferius requested, it would be interesting to see what happens if the offending phrase/script (and its corresponding .json file) are manually moved to somewhere AutoKey can't find them and AutoKey is restarted. If the problem is in the particular script/phrase, the behavior should disappear.

**Comment by CCmesch** [comment]
Hey, thanks for answering. When I remove the offending script/phrase, it just takes another one. First it had a date script I configured. After moving that away it took another one of my phrases. Here is the metadata (I've changed the phrase itself and the abbreviation, because it was my clear name):

```
{
    "usageCount": 0,
    "filter": {
        "isRecursive": false,
        "regex": null
    },
    "modes": [
        1
    ],
    "prompt": false,
    "hotkey": {
        "hotKey": null,
        "modifiers": []
    },
    "matchCase": false,
    "type": "phrase",
    "sendMode": "kb",
    "omitTrigger": false,
    "abbreviation": {
        "wordChars": "[\\w]",
        "backspace": true,
        "ignoreCase": true,
        "triggerInside": false,
        "abbreviations": [
            "+mph"
        ],
        "immediate": true
    },
    "showInTrayMenu": false,
    "description": "myphrase"
```

here is the meta data of the script:

```
{
    "store": {},
    "usageCount": 6,
    "filter": {
        "regex": null,
        "isRecursive": false
    },
    "type": "script",
    "description": "Insert Date",
    "omitTrigger": false,
    "hotkey": {
        "hotKey": null,
        "modifiers": []
    },
    "abbreviation": {
        "immediate": true,
        "triggerInside": false,
        "wordChars": "[\\w]",
        "ignoreCase": true,
        "abbreviations": [
            "+datum"
        ],
        "backspace": true
    },
    "modes": [
        1
    ],
    "showInTrayMenu": false,
    "prompt": false
}
```

**Comment by luziferius** [comment]
Those don’t look suspicious. I thought that there might be an empty abbreviation defined, like 
```
"abbreviations": [ "" ]
```
But that is not the case. Honestly, I have currently no idea what could lead to this situation.


Was it like this since you installed AutoKey or did it start some time later?

**Comment by luziferius** [comment]
Those don’t look suspicious. I thought that there might be an empty abbreviation defined, like 
```
"abbreviations": [ "" ]
```
But that is not the case. Honestly, I have currently no idea what could lead to this situation.


Was it like this since you installed AutoKey or did it start some time later?

**Comment by CCmesch** [comment]
Hi. Thanks for caring. I have several computers with autokey installed. It has been installed for some time and was working fine. It looks like it was after the normal system update (I do it via zypper) that is gradually crept into one by one of the computers. The issue is now in everyone of them.
But the problem is, at opensuse they update tons of packages, so it could be anything that causes the problem. I don't reckon I have a chance telling them and get help. And as there is the bug mentioned in autokey log (see above) I naturally came here. Can't anyone make anything out of this log error? 
Here is the offending bit again:
```
Ignored locking error in handle_keypress
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/autokey/service.py", line 206, in __tryReleaseLock
    self.configManager.lock.release()
RuntimeError: release unlocked lock
```
Thanks for helping!

**Comment by luziferius** [comment]
### RuntimeError
That thing was there for ages, at least 8 years. The original developer just silenced it.
I unmuted the message some time ago to make it visible and maybe fix it in the future.

I don't think that it'll cause these issues. I think this message is caused by a serialisation mutex that is released too often. There is a bug somewhere, but it does not cause deadlocks, at least.

### Your issue
- Ok, so it did not happen instantly after installation.
- The phrase meta-data you posted looks ok.
- They have in common that both have an abbreviation starting with `+`. Is that coincidence or a pattern?
- This seems to be not a common issue, otherwise there'd been more reports or questions in one of the support channels.(Google groups, irc, gitter chat...)
- At some point, autokey started to expand stuff on each keypress.

One thing: Look at the global autokey configuration file `autokey.json`, it is stored at `~/.config/autokey`. Please post the content. Censor personal data, if any, as you deem appropriate.

Another possibility to try: Backup all autokey files, delete them and start from scratch. If the issue goes away, it is related to your data files. If not, it may be a bug introduced by the packaging @ OpenSuse.
Then I'll have to verify their packaging.

**Comment by luziferius** [comment]
@CCmesch I just got an idea: Download a source code archive of the current version (and additionally the version equal to the one from the repositories) from the [Github releases page](https://github.com/autokey/autokey/releases).
Then
- Extract the source archives
- Exit any running AutoKey instance. Verify it is not running.
- Try the older version first, then the newer one:
  - Start the downloaded version directly from the extracted source archive. See here for details: https://github.com/autokey/autokey#zero-installation-method

Possible results:
- If the older version works fine, the packaging @ OpenSuse is at fault here.
- If the newer version works, but the older one not, consider it fixed by accident.
- If both show the same issue, it is either some system state or your user config.

**Comment by luziferius** [comment]
Seems related to the newest issue #254 
It is caused by the `Ignore case` option together with the `trigger immediately` setting in the abbreviation settings.

**Comment by luziferius** [comment]
This may be fixed in the just released version 0.95.6. Can you try the new release?

**Comment by CCmesch** [comment]
Thanks luziferus for the comments! I have tried your method of installing the latest package via "pip install". It seems to be working perfectly as yet. Version 0.95.6.

Thanks very much!

I'll close this now. If I experience more problems, I'll try opening it again.

**Comment by luziferius** [comment]
You were running on stock 0.95.4, right? Not git master, but just the plain release, tagged v0.95.4?

If so, the bug #254 that I introduced in 0.95.5 was also present in prior versions, but wasn’t that prominently triggered.

**Comment by CCmesch** [comment]
Yes, I have already said that I installed it via package manager zypper, in  v0.95.4
But I don't see why that one should be different from any other v0.95.4 version.
But anyway, the error is solved in 0.95.6, so all is well.

Cheerio!
CCmesch
